File: ffi/ts/src/keys/publicKey.ts
import {
  calcNonce,
  castToPubKey,
  freeObj,
  genRandomPublicKey,
  getPublicKeyPoint,
  pointToPublicKey,
  scalarToPubKey,
} from '../blsct'

import { ManagedObj } from '../managedObj'
import { Point } from '../point'
import { Scalar } from '../scalar'

/** Represents an element in the BLS12-381 G1 curve group that is used as a public key.
 *
 * Examples:
 * ```ts
 * const { PublicKey, Point, Scalar } = require('navio-blsct')
 * const s123 = new Scalar(123)
 * const pk123 = PublicKey.fromScalar(s123)
 * const s234 = new Scalar(234)
 * const pk234 = PublicKey.fromScalar(s234)
 * pk123.equals(pk234) // false
 * pk123.equals(pk123) // true
 * const p = Point.random()
 * const pk = PublicKey.fromPoint(p)
 * pk.getPoint()
 * const vk = Scalar.random()
 * pk.generateNonce(vk)
 * const ser = pk.serialize()
 * const deser = PublicKey.deserialize(ser)
 * ser == deser.serialize()
 * ```
 */
export class PublicKey extends ManagedObj {
  /** Constructs a new `PublicKey` instance.
   * - If no parameter is provided, a random public key is generated.
   */
  constructor(obj?: any) {
    if (typeof obj === 'object') {
      super(obj)
    } else {
      const rv = genRandomPublicKey()
      if (rv.result !== 0) {
        throw new Error('Failed to generate random PublicKey')
      }
      super(rv.value)
    }
  }

  /** Return the underlying point of the public key
   * @return {Point} The underlying point of the public key.
   */
  getPoint(): Point {
    const blsctPoint = getPublicKeyPoint(this.value())
    return Point.fromObj(blsctPoint)
  }

  /** Returns a random public key.
   * @returns A new random `PublicKey`.
   */
  static random(): PublicKey {
    const rv = genRandomPublicKey()
    const pk = PublicKey.fromObj(rv.value)
    freeObj(rv)
    return pk
  }

  /** Creates a `PublicKey` from a `Point` object.
   * @param point - The `Point` object to convert.
   */
  static fromPoint(point: Point): PublicKey {
    const blsctPubKey = pointToPublicKey(point.value())
    return PublicKey.fromObj(blsctPubKey)
  }

  /** Creates a `PublicKey` from a `Scalar` object.
   * @param scalar - The `Scalar` object to convert.
   */
  static fromScalar(scalar: Scalar): PublicKey {
    const blsctPubKey = scalarToPubKey(scalar.value())
    return PublicKey.fromObj(blsctPubKey)
  }

  /** Generates a nonce from this `PublicKey` using a `Scalar` view key.
   *
   * @param viewKey - The view key.
   * @returns A new `PublicKey` that represents the nonce.
   */
  generateNonce(
    viewKey: Scalar
  ): PublicKey {
    const blsctNonce = calcNonce(
      this.value(),
      viewKey.value()
    )
    return PublicKey.fromObj(blsctNonce)
  }

  /** Returns if the provied public key is equal to this public key.
   * @param other - The public key to compare with.
   * @returns `true` if the public keys are equal, `false` otherwise.
   */
  equals(other: PublicKey): boolean {
    return this.getPoint().equals(other.getPoint())
  }

  override value(): any {
    return castToPubKey(this.obj)
  }

  override serialize(): string {
    return this.getPoint().serialize()
  }

  /** Deserializes a hexadecimal string into a `PublicKey` instance.
   * @param hex - The hexadecimal string to convert.
   * @returns The `PublicKey` instance represented by the input string.
   */
  static deserialize(
    this: new (obj: any) => PublicKey,
    hex: string
  ): PublicKey {
    const p = Point.deserialize(hex)
    return PublicKey.fromPoint(p)
  }
}

File: ffi/rust/src/keys/public_key.rs
use crate::{
  blsct_obj::{BlsctObj, self},
  blsct_serde::BlsctSerde, 
  ffi::{
    BLSCT_FAILURE,
    BlsctPubKey,
    BlsctRetVal,
    calc_nonce,
    err_bool,
    gen_random_public_key,
    get_public_key_point,
    point_to_public_key,
    PUBLIC_KEY_SIZE,
    scalar_to_pub_key,
    serialize_point,
  },
  macros::{
    impl_clone,
    impl_display,
    impl_from_retval,
    impl_value,
  },
  point::Point,
  scalar::Scalar,
  util::{
    c_hex_str_to_array,
    build_succ_blsct_ret_val,
  },
};
use std::ffi::c_char;
use serde::{Deserialize, Serialize};

#[derive(PartialEq, Eq, Debug, Deserialize, Serialize)]
pub struct PublicKey {
  obj: BlsctObj<PublicKey, BlsctPubKey>,
}

impl_from_retval!(PublicKey);
impl_display!(PublicKey);
impl_clone!(PublicKey);

impl PublicKey {
  pub fn random<'a>() -> Result<Self, blsct_obj::Error<'a>> {
    Self::from_retval(unsafe { gen_random_public_key() })
  }

  pub fn generate_nonce(&self, view_key: &Scalar) -> Self {
    let blsct_point = unsafe {
      calc_nonce(self.value(), view_key.value())
    };
    let obj = BlsctObj::from_c_obj(blsct_point);
    let point: Point = obj.into();
    (&point).into()
  }

  pub fn get_point(&self) -> Point {
    let blsct_point = unsafe { get_public_key_point(self.value()) };
    let obj = BlsctObj::from_c_obj(blsct_point);
    obj.into()
  }

  impl_value!(BlsctPubKey);
}

impl BlsctSerde for PublicKey {
  unsafe fn serialize(ptr: *const u8, _: usize) -> *const i8 {
    // serialize the PublicKey as a Point
    let point = get_public_key_point(ptr as *const BlsctPubKey);
    serialize_point(point)
  }

  unsafe fn deserialize(hex: *const c_char) -> *mut BlsctRetVal {
    // since hex is a serialized Point, convert it back to BlsctPoint 
    match c_hex_str_to_array(hex) {
      Ok(buf) => {
        let blsct_pub_key = unsafe { point_to_public_key(&buf) };
        match build_succ_blsct_ret_val::<PUBLIC_KEY_SIZE>(blsct_pub_key as *const u8) {
          Ok(rv) => rv,
          Err(_) => err_bool(BLSCT_FAILURE),
        }
      },
      Err(_) => {
        err_bool(BLSCT_FAILURE)
      },
    }
  }
}

impl From<&PublicKey> for Point {
  fn from(pub_key: &PublicKey) -> Point {
    let blsct_point = unsafe { get_public_key_point(pub_key.value()) };
    let obj = BlsctObj::from_c_obj(blsct_point);
    obj.into()
  }
}

impl From<BlsctObj<PublicKey, BlsctPubKey>> for PublicKey {
  fn from(obj: BlsctObj<PublicKey, BlsctPubKey>) -> PublicKey {
    PublicKey { obj }
  }
}

impl From<&Point> for PublicKey {
  fn from(point: &Point) -> PublicKey {
    let blsct_pub_key = unsafe { point_to_public_key(point.value()) };
    let obj = BlsctObj::from_c_obj(blsct_pub_key);
    obj.into()
  }
}

impl From<&Scalar> for PublicKey {
  fn from(scalar: &Scalar) -> PublicKey {
    let blsct_pub_key = unsafe { scalar_to_pub_key(scalar.value()) };
    let obj = BlsctObj::from_c_obj(blsct_pub_key);
    obj.into()
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::{
    initializer::init,
    keys::child_key::ChildKey,
  };

  #[test]
  fn test_random() {
    init();
    let _: PublicKey = PublicKey::random().unwrap();
  }

  #[test]
  fn test_generate_nonce() {
    init();
    let pub_key = PublicKey::random().unwrap();
    let child_key = ChildKey::random().unwrap();
    let view_key = child_key.to_tx_key().to_view_key();
    let _: PublicKey = pub_key.generate_nonce(&view_key);
  }

  #[test]
  fn test_from_scalar() {
    init();
    let scalar = Scalar::random().unwrap();
    let _: PublicKey = (&scalar).into();
  }

  #[test]
  fn test_from_point() {
    init();
    let point = Point::random().unwrap();
    let _: PublicKey = (&point).into();
  }

  #[test]
  fn test_to_point() {
    init();
    let pub_key = PublicKey::random().unwrap();
    let _: Point = (&pub_key).into();
  }

  #[test]
  fn test_eq() {
    init();
    let (a, b) = {
      loop {
        let a = PublicKey::random().unwrap();
        let b = PublicKey::random().unwrap();
        if a != b {
          break (a, b);
        }
      }
    };
    assert!(a == a);
    assert!(a != b);
    assert!(b != a);
    assert!(b == b);
  }

  #[test]
  fn test_deser() {
    init();
    let a = PublicKey::random().unwrap();
    let hex = bincode::serialize(&a).unwrap();
    let b = bincode::deserialize::<PublicKey>(&hex).unwrap();

    assert_eq!(a, b);
  }
}

